### ISSUE TITLE
OCI helm chart with dependencies not picking helm repo secret

### ISSUE THREAD

**Issue by singhravinderlko** [argoproj/argo-cd#8266]
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [y] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [y ] I've included steps to reproduce the bug.
* [ y] I've pasted the output of `argocd version`.

**Describe the bug**

i am using gcp artifact helm repository for storing OCI helm chart. when trying to create a application using oci helm chart with dependencies, it is failing during dependencies resoluton with authentication related errors(error pasted below). while if a just create a argocd app with no helm chart dependencies, it works fine. i am using credential template for authenticating to private GCP artifact helm repository. 

rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: could not download oci://<helm OCI repo path>: failed to authorize: failed to fetch anonymous token: unexpected status: 403 Forbidden

**To Reproduce**
1- create argocd application using below yaml which is trying to deploy chart.yaml which has been kept on github repo and in dependencies i am referring to gcp artifact repository which is a OCI helm repo-
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: prismatesting
  namespace: argocd
spec:
  project: devops
  source:
    repoURL: https://<github-repo-url>
    targetRevision: HEAD
    path: clusters/devops/twistlock-defender
    helm:
      valueFiles:
      - values.yaml
  destination:
    name: devops-clusterA
    namespace: twistlock
  syncPolicy:
    automated:
      prune: true
```
```
apiVersion: v1
description: Twistlock Defender Daemon Set Helm Chart
name: twistlock-defender
version: 21.08.520
dependencies:
- name: twistlock-defender
  version: 21.08.525
  repository: 'oci://<oci helm repo url>'
```
2- application crd gets deployed but fails to apply the dependencies chart with following error:

rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: could not download oci://<oci helm repo path>: failed to authorize: failed to fetch anonymous token: unexpected status: 403 Forbidden

**Expected behavior**

application should have been able to apply dependencies chart without authentication failure.

**Screenshots**

**Version**

```shell
Paste the output from `argocd version` here.
argocd: v2.0.5+4c94d88
  BuildDate: 2021-07-22T20:52:31Z
  GitCommit: 4c94d886f56bcb2f9d5b3251fdc049c2d1354b88
  GitTreeState: clean
  GoVersion: go1.16
  Compiler: gc
  Platform: linux/amd64
argocd-server: v2.2.2+03b17e0
```

**Logs**

```
rpc error: code = Unknown desc = Manifest generation error (cached): `helm dependency build` failed exit status 1: Error: could not download oci://<repo path>: failed to authorize: failed to fetch anonymous token: unexpected status: 403 Forbidden
```

**Comment by ocraviotto** [comment]
@singhravinderlko Support for credential templates for oci dependencies was introduced in #6027 and then improved in #6994
I came here to report a pending issue when using credential templates only with oci private registries (without a repository matching), but the error is different in those cases (more below). 

Could you please confirm the same credentials in the repocreds secret work if you use them to locally when logining with helm?
```sh
kubectl get secrets [your-creds-secret] -o json | jq -r '.data.password' | base64 -d | helm registry login --username $(kubectl get secrets [your-creds-secret] -o json | jq -r '.data.username' | base64 -d) --password-stdin [your registry URL]
``` 
Remember to either use helm 3.8 or export HELM_EXPERIMENTAL_OCI=1 

For my case, I'd like to know: are you only using a credentials template (secret of type repocreds) or do you also define a repository?

As mentioned, I'm asking because I came for a similar issue, but I have identified the problem and though related, it's a bit different from yours:
when setting credentials templates that match a registry but that are not given as a repository (`repocreds` instead), the login fails because Argo CD [passes the schema `oci://` as the registry URL](https://github.com/argoproj/argo-cd/blob/v2.3.0-rc1/reposerver/repository/repository.go#L487-L511), but helm (using oras-go) will fail because it does not support the oci:// schema, and so login command will fail with something like `Error: Get "https://oci/v2/": dial tcp: lookup oci on 127.0.0.53:53: server misbehaving`

Anyway, I'll create another issue with these details, was simply wondering what you issue was exactly in case it was related. Will try to hep if I can regardless.

**Comment by jerousseau** [comment]
@ocraviotto I couldn't find your issue so I created one : #8563

**Comment by chenfli** [comment]
Encounter this as well on Azure Container Regisry 
ArgoCD v2.3.3+07ac038
Helm v3.8.0+gd141386

**Comment by turnerno8** [comment]
We are also seeing this issue with the 2.4.0-rc1

Is there any workaround or expected fix?

{
    "Version": "v2.4.0-rc1+d8f845a",,
    "Platform": "linux/amd64",
    "KustomizeVersion": "v4.4.1 2021-11-11T23:36:27Z",
    "HelmVersion": "v3.8.1+g5cb9af4",
    "KubectlVersion": "v0.23.1",
    "JsonnetVersion": "v0.18.0"
}

**Comment by turnerno8** [comment]
Sorry- My issue was with the project allowed repos. It worked with rc-1

**Comment by blakepettersson** [comment]
This should be addressed with #11327.
